### PR TITLE
Function hiding inner product schemes (basic and multi-client)

### PR DESCRIPTION
**Pull request by tilenmarc**
This pull request implements two function hiding inner product schemes from papers:
- [Function-Hiding Inner Product Encryption is Practical](https://eprint.iacr.org/2016/440.pdf)
- [Full-Hiding (Unbounded) Multi-Input Inner Product Functional Encryption from the k-Linear Assumption](https://eprint.iacr.org/2018/061.pdf)
The implementation reflects the implementation in GoFE, in particular the multi-client scheme is slightly modified from the presentation in the paper, to achieve better performance. In addition to the implementation of the schemes:
- Gaussian elimination that was used in ABE schemes in moved to data/mat.c file
- Additional functions using Gaussian elimination are implemented: calculating an inverse and determinant of a matrix.
- Two new functions with vectors of elements of elliptic curve are implemented in data/vec_curve.c

Note that this PR should be merged after PR with decentralized functions is merged.

### CODE REVIEW

File: include/cifer/data/mat.h [hunk 1]
```diff
@@ -201,4 +201,57 @@ cfe_error cfe_mat_inverse_mod(cfe_mat *inverse_mat, cfe_mat *m, mpz_t mod);
...
+ *
+ * @param res A pointer to an initialized matrix where the result will be saved
+ * @param det Determinant will be saved here; if not needed this should be NULL
+ * @param mat A pointer to a matrix
+ * @param p Modulus for the computations
+ */
+cfe_error cfe_mat_inverse_mod_gauss(cfe_mat *res, mpz_t det, cfe_mat *m, mpz_t p);
+
+/**
+ * cfe_mat_inverse_mod_gauss calculates the determinant of a matrix over field
```
Comment (miha-stopar) [inline include/cifer/data/mat.h]: Fix function name in a comment.

File: include/cifer/data/mat.h [hunk 2]
```diff
@@ -201,4 +201,57 @@ cfe_error cfe_mat_inverse_mod(cfe_mat *inverse_mat, cfe_mat *m, mpz_t mod);
...
+ * Z_p, using Gaussian elimination. The latter is faster than the naive (analytic)
+ * algorithm.
+ *
+ * @param res The result will be saved
+ * @param mat A pointer to a matrix
+ * @param p Modulus for the computations
+ */
+void cfe_mat_determinant_gauss(mpz_t det, cfe_mat *m, mpz_t p);
```
Comment (miha-stopar) [inline include/cifer/data/mat.h]: Fix name in a comment.

File: include/cifer/innerprod/fullysec/fhipe.h [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+ * @param sec_key A pointer to a cfe_fhipe_sec_key struct (the master secret key will
+ * be stored here)
+ * @param c A pointer to an instance of the scheme (*initialized* cfe_fhipe
+ * struct)
+ * @return Error code
+ */
+cfe_error cfe_fhipe_generate_master_key(cfe_fhipe_sec_key *sec_key, cfe_fhipe *c);
+
+/**
+ * cfe_fhipe_FE_key represents a functional encryption key in fhipe scheme.
```
Comment (miha-stopar) [inline include/cifer/innerprod/fullysec/fhipe.h]: Fix name.

**Comment by tilenmarc** [comment]
Sharp eye. I changed the names and also remembered that readme should be updated. I will merge.

**Pull request merged**